### PR TITLE
Support QTTabBar

### PR DESCRIPTION
**Pull request by onesounds**
## What's the PR
- Support for opening tab in exist window when using qttabbar.
- Fix https://github.com/Flow-Launcher/Flow.Launcher/issues/2790

## Detail 
- Added a QtTabBar in "File Manager" entry. 
- For new installations, the qttabbar setting is listed.
- Existing users are not listing. but if user add and select "qttabbar" (in name. case-insensitive), it will behave as a qttabbar.

## Test Cases
- [x] opening in tabs in qttabbar.
- [x] Doesn't affect existing user settings.

### CODE REVIEW

File: Flow.Launcher/PublicAPIInstance.cs [hunk 1]
```diff
@@ -228,21 +229,39 @@ public void SaveJsonStorage<T>(T settings) where T : new()
 
         public void OpenDirectory(string DirectoryPath, string FileNameOrFilePath = null)
         {
-            using var explorer = new Process();
+            var customExplorerList = _settingsVM.Settings.CustomExplorerList;
```
[the remark this answers is not in the document]
Comment (onesounds) [reply Flow.Launcher/PublicAPIInstance.cs:232]: OK.

File: Flow.Launcher/PublicAPIInstance.cs [hunk 2]
```diff
@@ -228,21 +229,39 @@ public void SaveJsonStorage<T>(T settings) where T : new()
...
-                UseShellExecute = true,
-                Arguments = FileNameOrFilePath is null
-                    ? explorerInfo.DirectoryArgument.Replace("%d", DirectoryPath)
-                    : explorerInfo.FileArgument
-                        .Replace("%d", DirectoryPath)
-                        .Replace("%f",
-                            Path.IsPathRooted(FileNameOrFilePath) ? FileNameOrFilePath : Path.Combine(DirectoryPath, FileNameOrFilePath)
-                        )
-            };
-            explorer.Start();
```
Comment (jjw24) [inline Flow.Launcher/PublicAPIInstance.cs:245]: Does this existing process start logic not work for QTTabBar setup?

**Review by jjw24 — CHANGES_REQUESTED** [review]
We need to handle the scenario where the new OTTabBar entry should show up on existing user data, because the way the code loads the existing user settings it won't load the new entry in unless it's a fresh install.

**Comment by onesounds** [comment]
> We need to handle the scenario where the new OTTabBar entry should show up on existing user data, because the way the code loads the existing user settings it won't load the new entry in unless it's a fresh install.

I agree, but I don't know how to do it. I considered adding a separate setting for this item for existing user, but there was no proper location. (And since qtttabbar is not being maintained. so I didn't want to add a setting menu for this.)

Let me ask chatgpt...

**Comment by onesounds** [comment]
> We need to handle the scenario where the new OTTabBar entry should show up on existing user data, because the way the code loads the existing user settings it won't load the new entry in unless it's a fresh install. In this case, the qttabbar is an item that cannot be deleted.

Now, there are three ways to support this. I can work with 1 or 2. 
1. Each time the user launches, we check the user's settings to see if the "qttabbar" entry exists, and if not, we add it to the list.

2. Add qttabbar and explorer as user-immutable constant entries and adjust the list.
In this case, we either have to risk messing with existing users' FileManager settings(settings.CustomExplorerIndex), or we have to reset it and change it to store and recognize by "name" (string). Also, in this case, qttabbar, which should be lower in the support list, is occupying the second list.

3.If the user doesn't have qttabbar setting, it should be added automatically, and if it is already added, it should not be added again if the user subsequently deletes it.  I don't think I can do this, and while it's probably possible, it results in an incredibly complex structure. Is there any other easy way to do this? 

This is entirely down to me structuring it incorrectly at the beginning, However, I don't know if it's worth adding such a structure for this utility. Given the importance of qttabbar support, I think my PR state is good enough. I can't do 3.  After reviewing all 3, I still think what I proposed is the right way to go, but if you want to do 1 and 2 to add a qttabbar item to the list that almost all existing users will never use, that's fine. Just pick one and let me know.

**Comment by onesounds** [comment]
I'm closing that PR because I don't think my plan is clear and it's causing confusion. If I come up with a new idea, I'll write a new one, or others can make one.

**Comment by jjw24** [comment]
> I'm closing that PR because I don't think my plan is clear and it's causing confusion. If I come up with a new idea, I'll write a new one, or others can make one.

Thanks for helping to investigate how to get QTTabBar integrated. We just need to do a slight tweaking that will have less complexity, please see my commit. The commit will allow QTTabBar be added as a custom profile, which means there is no need to provide backwards support for a new default profile. We can add it like this:


image

**Comment by jjw24** [comment]
@onesounds could you please help make the change to allow the 'Arg For Folder' and 'Arg For File' fields to be blank.

**Comment by onesounds** [comment]
> 'Arg For Folder' and 'Arg For File'

@jjw24 Done.

**Review by jjw24 — APPROVED** [review]

**Pull request merged**